dbms/src/Interpreters/InterpreterSelectQuery.cpp

```diff
@@ -75,6 +76,39 @@ namespace ErrorCodes
     extern const int INVALID_LIMIT_EXPRESSION;
 }
 
+namespace
+{
+
+/// Assumes `storage` is set and the table filter is not empty.
+String generateFilterActions(ExpressionActionsPtr & actions, const StoragePtr & storage, const Context & context, const Names & prerequisite_columns = {})
+{
+    ParserSelectQuery parser;
+    const auto & db_name = storage->getDatabaseName();
+    const auto & table_name = storage->getTableName();
+    const auto & filter_str = context.getUserProperty(db_name, table_name, "filter");
+
+    /// Keep columns that are required after the filter actions.
+    String columns;
+    for (const auto & column_name : prerequisite_columns)
+    {
+        columns += column_name + ',';
+    }
+
+    /// Using separate expression analyzer to prevent any possible alias injection
+    String query_str = "select " + columns + filter_str + " from " + db_name + "." + table_name + " where " + filter_str;
+    auto query_ast = parseQuery(parser, query_str, 0);
+
+    auto syntax_result = SyntaxAnalyzer(context).analyze(query_ast, storage->getColumns().getAllPhysical());
```

Comment:
I think we can just pass filter_str as ast here and prerequisite_columns as additional_source_columns. But if it's difficult, than current solution is ok.

Reply:
Discussed in person - decided to keep the `select` expression for now.